File: ios/MockStarket/Core/Networking/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case invalidURL
    case invalidResponse
    case httpError(Int, String)
    case decodingError(Error)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid URL"
        case .invalidResponse: return "Invalid response"
        case .httpError(let code, let message): return "HTTP \(code): \(message)"
        case .decodingError(let error): return "Decoding error: \(error.localizedDescription)"
        case .networkError(let error): return "Network error: \(error.localizedDescription)"
        }
    }
}

enum APIEndpoint {
    // Auth
    case register
    case createGuest
    case getMe
    case updateMe
    case deleteMe

    // Stocks
    case listStocks
    case getStock(ticker: String)
    case getStockHistory(ticker: String, interval: String)
    case getETFHoldings(ticker: String)
    case marketSummary

    // Trading
    case executeTrade
    case getTradeHistory(limit: Int, offset: Int)

    // Orders
    case createOrder
    case listOrders
    case cancelOrder(id: UUID)

    // Portfolio
    case getPortfolio
    case getPortfolioHistory

    // Leaderboard
    case getLeaderboard(period: String)

    // Alerts
    case createAlert
    case listAlerts
    case deleteAlert(id: UUID)

    // Achievements
    case listAchievements
    case getMyAchievements

    // Watchlist
    case getWatchlist
    case addToWatchlist
    case removeFromWatchlist(ticker: String)

    // Challenges
    case getTodaysChallenge
    case checkChallenge
    case claimChallenge(id: UUID)

    // System
    case health

    var path: String {
        switch self {
        case .register: return "/api/v1/auth/register"
        case .createGuest: return "/api/v1/auth/guest"
        case .getMe, .updateMe, .deleteMe: return "/api/v1/auth/me"
        case .listStocks: return "/api/v1/stocks"
        case .getStock(let ticker): return "/api/v1/stocks/\(ticker)"
        case .getStockHistory(let ticker, let interval): return "/api/v1/stocks/\(ticker)/history?interval=\(interval)"
        case .getETFHoldings(let ticker): return "/api/v1/stocks/\(ticker)/holdings"
        case .marketSummary: return "/api/v1/stocks/market-summary"
        case .executeTrade, .getTradeHistory: return "/api/v1/trades"
        case .createOrder, .listOrders: return "/api/v1/orders"
        case .cancelOrder(let id): return "/api/v1/orders/\(id)"
        case .getPortfolio: return "/api/v1/portfolio"
        case .getPortfolioHistory: return "/api/v1/portfolio/history"
        case .getLeaderboard(let period): return "/api/v1/leaderboard?period=\(period)"
        case .createAlert, .listAlerts: return "/api/v1/alerts"
        case .deleteAlert(let id): return "/api/v1/alerts/\(id)"
        case .listAchievements: return "/api/v1/achievements"
        case .getMyAchievements: return "/api/v1/achievements/me"
        case .getWatchlist, .addToWatchlist: return "/api/v1/watchlist"
        case .removeFromWatchlist(let ticker): return "/api/v1/watchlist/\(ticker)"
        case .getTodaysChallenge: return "/api/v1/challenges/today"
        case .checkChallenge: return "/api/v1/challenges/check"
        case .claimChallenge(let id): return "/api/v1/challenges/\(id)/claim"
        case .health: return "/api/v1/system/health"
        }
    }

    var method: String {
        switch self {
        case .register, .createGuest, .executeTrade, .createOrder, .createAlert, .addToWatchlist, .checkChallenge, .claimChallenge:
            return "POST"
        case .updateMe:
            return "PUT"
        case .deleteMe, .cancelOrder, .deleteAlert, .removeFromWatchlist:
            return "DELETE"
        default:
            return "GET"
        }
    }
}

actor APIClient {
    static let shared = APIClient()

    private let session: URLSession
    private let decoder: JSONDecoder
    private var baseURL: String

    init(baseURL: String = "http://localhost:8080") {
        self.baseURL = baseURL
        self.session = URLSession.shared

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        self.decoder = decoder
    }

    func setBaseURL(_ url: String) {
        self.baseURL = url
    }

    func request<T: Decodable>(_ endpoint: APIEndpoint, token: String? = nil, body: Encodable? = nil) async throws -> T {
        guard let url = URL(string: baseURL + endpoint.path) else {
            throw APIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = endpoint.method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        if let body {
            request.httpBody = try JSONEncoder().encode(body)
        }

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw APIError.networkError(error)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            let errorMessage = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw APIError.httpError(httpResponse.statusCode, errorMessage)
        }

        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw APIError.decodingError(error)
        }
    }
}

File: web/src/lib/api-client.ts
const API_URL = process.env.NEXT_PUBLIC_API_URL || '';

class APIClient {
  private baseURL: string;
  private token: string | null = null;

  constructor(baseURL: string = API_URL) {
    this.baseURL = baseURL;
  }

  setToken(token: string | null) {
    this.token = token;
  }

  async request<T>(path: string, options: RequestInit = {}): Promise<T> {
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
      ...options.headers,
    };

    if (this.token) {
      (headers as Record<string, string>)['Authorization'] = `Bearer ${this.token}`;
    }

    const response = await fetch(`${this.baseURL}${path}`, {
      ...options,
      headers,
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`HTTP ${response.status}: ${error}`);
    }

    return response.json();
  }

  // Auth
  register(displayName: string, isGuest: boolean) {
    return this.request<any>('/api/v1/auth/register', {
      method: 'POST',
      body: JSON.stringify({ display_name: displayName, is_guest: isGuest }),
    });
  }

  createGuest() {
    return this.request<any>('/api/v1/auth/guest', { method: 'POST' });
  }

  getMe() {
    return this.request<any>('/api/v1/auth/me');
  }

  // Stocks
  getStocks() {
    return this.request<any[]>('/api/v1/stocks');
  }

  getStock(ticker: string) {
    return this.request<any>(`/api/v1/stocks/${ticker}`);
  }

  getStockHistory(ticker: string, interval: string = '1m') {
    return this.request<any[]>(`/api/v1/stocks/${ticker}/history?interval=${interval}`);
  }

  getETFHoldings(ticker: string) {
    return this.request<any[]>(`/api/v1/stocks/${ticker}/holdings`);
  }

  getMarketSummary() {
    return this.request<any>('/api/v1/stocks/market-summary');
  }

  // Trading
  executeTrade(ticker: string, side: string, shares: number) {
    return this.request<any>('/api/v1/trades', {
      method: 'POST',
      body: JSON.stringify({ ticker, side, shares }),
    });
  }

  getTradeHistory(limit = 50, offset = 0) {
    return this.request<any[]>(`/api/v1/trades?limit=${limit}&offset=${offset}`);
  }

  // Portfolio
  getPortfolio() {
    return this.request<any>('/api/v1/portfolio');
  }

  // Leaderboard
  getLeaderboard(period: string = 'alltime') {
    return this.request<any[]>(`/api/v1/leaderboard?period=${period}`);
  }

  // Orders
  createOrder(ticker: string, side: string, orderType: string, shares: number, limitPrice?: string, stopPrice?: string) {
    return this.request<any>('/api/v1/orders', {
      method: 'POST',
      body: JSON.stringify({ ticker, side, order_type: orderType, shares, limit_price: limitPrice, stop_price: stopPrice }),
    });
  }

  getOrders() {
    return this.request<any[]>('/api/v1/orders');
  }

  cancelOrder(id: string) {
    return this.request<any>(`/api/v1/orders/${id}`, { method: 'DELETE' });
  }

  // Alerts
  createAlert(ticker: string, condition: string, targetPrice: string) {
    return this.request<any>('/api/v1/alerts', {
      method: 'POST',
      body: JSON.stringify({ ticker, condition, target_price: targetPrice }),
    });
  }

  getAlerts() {
    return this.request<any[]>('/api/v1/alerts');
  }

  deleteAlert(id: string) {
    return this.request<any>(`/api/v1/alerts/${id}`, { method: 'DELETE' });
  }

  // Achievements
  getAchievements() {
    return this.request<any[]>('/api/v1/achievements');
  }

  getMyAchievements() {
    return this.request<any[]>('/api/v1/achievements/me');
  }

  // Watchlist
  getWatchlist() {
    return this.request<string[]>('/api/v1/watchlist');
  }

  addToWatchlist(ticker: string) {
    return this.request<any>('/api/v1/watchlist', {
      method: 'POST',
      body: JSON.stringify({ ticker }),
    });
  }

  removeFromWatchlist(ticker: string) {
    return this.request<any>(`/api/v1/watchlist/${ticker}`, { method: 'DELETE' });
  }

  // Challenges
  getTodaysChallenge() {
    return this.request<any>('/api/v1/challenges/today');
  }

  checkChallenge() {
    return this.request<any>('/api/v1/challenges/check', { method: 'POST' });
  }

  claimChallenge(id: string) {
    return this.request<any>(`/api/v1/challenges/${id}/claim`, { method: 'POST' });
  }

  // Portfolio History
  getPortfolioHistory(limit = 100) {
    return this.request<any[]>(`/api/v1/portfolio/history?limit=${limit}`);
  }

  // Options
  getOptionChain(ticker: string, expiration?: string) {
    const params = expiration ? `?expiration=${expiration}` : '';
    return this.request<any>(`/api/v1/stocks/${ticker}/options${params}`);
  }

  getOptionExpirations(ticker: string) {
    return this.request<string[]>(`/api/v1/stocks/${ticker}/options/expirations`);
  }

  getOptionContract(id: string) {
    return this.request<any>(`/api/v1/options/${id}`);
  }

  executeOptionsTrade(contractId: string, side: string, quantity: number) {
    return this.request<any>('/api/v1/options/trades', {
      method: 'POST',
      body: JSON.stringify({ contract_id: contractId, side, quantity }),
    });
  }

  getOptionsTradeHistory(limit = 50, offset = 0) {
    return this.request<any[]>(`/api/v1/options/trades?limit=${limit}&offset=${offset}`);
  }

  getOptionsPositions() {
    return this.request<any[]>('/api/v1/options/positions');
  }

  createOptionsOrder(contractId: string, side: string, orderType: string, quantity: number, limitPrice?: string) {
    return this.request<any>('/api/v1/options/orders', {
      method: 'POST',
      body: JSON.stringify({ contract_id: contractId, side, order_type: orderType, quantity, limit_price: limitPrice }),
    });
  }

  getOptionsOrders() {
    return this.request<any[]>('/api/v1/options/orders');
  }

  cancelOptionsOrder(id: string) {
    return this.request<any>(`/api/v1/options/orders/${id}`, { method: 'DELETE' });
  }
}

export const apiClient = new APIClient();
